portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java

```diff
@@ -100,4 +133,71 @@ private BlobKey getBlobKey(HttpServletRequest request, String formInputElementNa
 
     return blobKey;
   }
+
+  /**
+   * Blobstore stores files as binary data. This function retrieves the binary data stored at the
+   * BlobKey parameter.
+   *
+   * @param blobKey The key associated with the image whose binary data is retrieved.
+   * @return An byte array containing the binary data of the image associated with the blobKey.
+   * @throws IOException - If an output error occurs when writing bytes from the temp blobstore
+   *     buffer {@code b} to the output byte array {@code outputBytes}.
+   */
+  private byte[] getBlobBytes(BlobKey blobKey) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+    boolean continueReading = true;
+    while (continueReading) {
+      // end index is inclusive, so we have to subtract 1 to get fetchSize bytes
+      byte[] b =
+          blobstoreService.fetchData(blobKey, currentByteIndex, currentByteIndex + fetchSize - 1);
+      outputBytes.write(b);
+
+      // if we read fewer bytes than we requested, then we reached the end
+      if (b.length < fetchSize) {
+        continueReading = false;
+      }
+
+      currentByteIndex += fetchSize;
+    }
+
+    return outputBytes.toByteArray();
+  }
+
+  /**
+   * Uses the Google Cloud Vision API to generate location information for a landmark detected in
+   * the image represented by the binary data stored in imgBytes.
+   *
+   * @param imgBytes Binary data for the image that is being inspected for landmark detection.
+   * @return An landmark annotation object containing information such as name and coordinates for
+   *     the landmark detected in the image. Null if there are no landmarks detected or other errors
+   *     occur when obtaining the landmark information.
+   * @throws IOException - If an input or output error occurs when creating the {@code
+   *     ImageAnnotatorClient} object.
+   */
+  private List<EntityAnnotation> getLandmarkInfo(byte[] imgBytes) throws IOException {
+    ByteString byteString = ByteString.copyFrom(imgBytes);
+    Image image = Image.newBuilder().setContent(byteString).build();
+
+    Feature feature = Feature.newBuilder().setType(Feature.Type.LANDMARK_DETECTION).build();
+    AnnotateImageRequest request =
+        AnnotateImageRequest.newBuilder().addFeatures(feature).setImage(image).build();
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    requests.add(request);
+
+    ImageAnnotatorClient client = ImageAnnotatorClient.create();
+    BatchAnnotateImagesResponse batchResponse = client.batchAnnotateImages(requests);
+    client.close();
+
+    AnnotateImageResponse imageResponse = batchResponse.getResponsesList().get(0);
```

Comment:
Will index 0 always definitively exist? I'm not familiar with these libraries so I'm just asking in case this is potentially unsafe.

Reply:
Based on my testing (no image at all, one where landmark is not detected, and one where landmark was detected) there was never an error. I just now put some print statements in there to see what imageResponse is in the case of no landmark. It is in fact an empty AnnotateImageResponse object. So it should theoretically be okay to keep it as is.

Reply:
If you think it is wise to add a conditional there to check that and return an empty list or null, I can do that. Otherwise I will keep it as is.